manticore/utils/mappings.py

```diff
@@ -41,15 +41,31 @@ def mmap(fd, offset, size):
     prot = MMAP.PROT_READ | MMAP.PROT_WRITE
     flags = MMAP.MAP_PRIVATE
 
+    # When trying to map the contents of a file into memory, the offset must be a multiple of the
+    # page size. So we need to align it before passing it to mmap(). But doing so also increases
+    # the size of the memory area needed, so we need to account for that difference.
+    aligned_offset = offset & ~0xfff
+    size += offset - aligned_offset
+
     if size & 0xfff != 0:
         size = (size & ~0xfff) + 0x1000
-
     assert size > 0
 
-    result = mmap_function(0, size, prot, flags, fd, offset)
-    return ctypes.cast(result, ctypes.POINTER(ctypes.c_char))
+    result = mmap_function(0, size, prot, flags, fd, aligned_offset)
+
+    # Now when returning the pointer to the user, we need to skip the corrected offset so that the
+    # user doesn't end up with a pointer to an other region of the file than the one he requested.
+    return ctypes.cast(result + offset - aligned_offset, ctypes.POINTER(ctypes.c_char))
 
 
 def munmap(address, size):
-    result = munmap_function(address, size)
+    # When unmapping the memory area, we need to recover the pointer and the size that were given
+    # to mmap(). The pointer can be recovered by aligning it on the page size. The size needs to be
+    # increased by that difference to account for cases where the requested memory area was larger.
+    address = ctypes.cast(address, ctypes.c_void_p).value
+    aligned_address = address & ~0xfff
+    size += address - aligned_address
+    aligned_address = ctypes.cast(aligned_address, ctypes.POINTER(ctypes.c_char))
```

Comment:
Is the cast required here at all? I assume that since we call a glibc function the type doesn't really matter here?

Reply:
No it isn't. I've kept it to conform to the definition of `munmap_function.argtype` but I removed it in my latest commit.